azkaban-common/src/main/java/azkaban/executor/container/KubernetesContainerizedImpl.java

```diff
@@ -398,10 +403,88 @@ public void createContainer(final int executionId) throws ExecutorManagerExcepti
+    logger.info(String.format("Cleaning up containers older than %d days",
+        containerValidity.toDays()));
+    final Set<Integer> containersToDelete = getStaleContainers(containerValidity);
+    for (final int execId : containersToDelete) {
+      if (execId > 0) {
+        logger.info(String.format("Deleting stale pod %s and service %s", getPodName(execId),
+            getServiceName(execId)));
+        try {
+          deleteContainer(execId);
+        } catch (ExecutorManagerException e) {
+          logger.error(String.format("Unable to delete stale pod and service of exec-id %d",
+              execId), e.getMessage());
+        }
+      }
+    }
+  }
+
+  /**
+   * This method is used to fetch all stale pods in the current az cluster and namespace
+   * @param containerValidity container valid duration
+   * @return Set of stale container's execution ids
+   */
+  private Set<Integer> getStaleContainers(final Duration containerValidity) throws ExecutorManagerException {
+    try {
+      // Select pods only from current Azkaban cluster and namespace
+      final String label = CLUSTER_LABEL_NAME + "=" + this.clusterName + "," + LABEL_POSTFIX;
+      final V1PodList items= this.coreV1Api.listNamespacedPod(this.namespace, null,
+          null, null, null, label,
+          null, null, null, null);
+
+      // Get all execution ids of the pods whose age is older than Azkaban max flow running time
+      // (e.g. 10 days)
+      final long validStartTimeStamp = System.currentTimeMillis() - containerValidity.toMillis();
+      return getExecutionIdFromPodList(items, validStartTimeStamp);
+    } catch (ApiException ae) {
+      logger.error(String.format("Unable to fetch stale pods in %s.", this.clusterName),
+          ae.getResponseBody());
+      throw new ExecutorManagerException(ae);
+    }
+  }
+
+  /**
+   * Obtain a set of execution ids from stale pod list
+   * @param podList
+   * @param validStartTimeStamp
+   * @return
+   */
+  @VisibleForTesting
+  Set<Integer> getExecutionIdFromPodList (final V1PodList podList, final long validStartTimeStamp) {
+    final Set<Integer> staleContainerExecIdSet = new HashSet<>();
+    for (final V1Pod pod: podList.getItems()) {
+      if (pod.getMetadata().getCreationTimestamp().getMillis() < validStartTimeStamp) {
+        final String execIdLabel =
+            pod.getMetadata().getLabels().getOrDefault(EXECUTION_ID_LABEL_NAME,
+            EXECUTION_ID_LABEL_PREFIX + DEFAULT_EXECUTION_ID);
+        try {
+          final String execId = execIdLabel.substring(execIdLabel.indexOf(
+              EXECUTION_ID_LABEL_PREFIX) + EXECUTION_ID_LABEL_PREFIX.length());
+          staleContainerExecIdSet.add(Integer.valueOf(execId));
+        } catch (Exception e) {
+          logger.error(String.format("Unable to retrieve execution id from pod %s",
+              pod.getMetadata().getName()), e.getMessage());
+        }
+      }
     }
```

Comment:
Thanks for encapsulating and testing this logic!
A few additional suggestions:
-`pod.getMetadata()` is called 3 times, consider calling once and reusing the result.

-I don't think `execIdLabel.indexOf()` call is needed as index should always be zero. 
`execIdLabel.substring(EXECUTION_ID_LABEL_PREFIX.length());` should be sufficient.

-this method skips executions that have started recently, please also consider exclusion of pods without an execution id (excluding execid--1 from the resulting list). 

-consider adding `s` in the method name: getExecutionId_**s**_FromPodList

Reply:
Used podMetadata instead of calling `pod.getMetadata()` 3 times. Changed the function name to `getExecutionIdsFromPodList`
exec-id == -1 is excluded in the `deleteAgedContainers`